jekyll/_posts/2021-09-25-version-160-released.md

```diff
@@ -0,0 +1,891 @@
+  in both `rst2html` (as before) as well as common tools rendering RST directly (e.g. GitHub). This is done by
+  adding the `default-role:: code` directive inside the RST file (which is now handled by `rst2html`).
+- Source+Edit links now appear on top of every docgen'd page when
+  `nim doc --git.url:url ...` is given.
+- Latex doc generation is revised: output `.tex` files should be compiled
+  by `xelatex` (not by `pdflatex` as before). Now default Latex settings
+  provide support for Unicode and do better job for avoiding margin overflows.
+- The RST parser now supports footnotes, citations, admonitions, short style references with symbols.
+- The RST parser now supports Markdown table syntax.
+  Known limitations:
+  - cell alignment is not supported, i.e. alignment annotations in a delimiter
+    row (`:---`, `:--:`, `---:`) are ignored,
+  - every table row must start with `|`, e.g. `| cell 1 | cell 2 |`.
+- Implemented `doc2tex` compiler command which converts documentation in
+  `.nim` files to Latex.
+- docgen now supports syntax highlighting for inline code.
+- docgen now supports same line doc comments:
+```nim
+func fn*(a: int): int = 42  ## Doc comment
+```
+- docgen now renders deprecated and other pragmas.
+- `runnableExamples` now works with templates and nested templates.
+- `runnableExamples: "-r:off"` now works for examples that should compile but not run. 
+- `runnableExamples` now renders code verbatim, and produces correct code in all cases.
+- docgen now shows correct, canonical import paths in docs.
+- docgen now shows all routines in sidebar, and the proc signature is now shown in sidebar.
+
+
+## Effects and checks
+- Significant improvement to error messages involving effect mismatches
+- There is a new `cast` section `{.cast(uncheckedAssign).}: body` that disables some
+  compiler checks regarding `case objects`. This allows serialization libraries
+  to avoid ugly, non-portable solutions. See RFC [#407](https://github.com/nim-lang/RFCs/issues/407).
+  for more details.
+
+
+## Tools
+- Major improvements to `nimgrep`, see PR [#15612
+](https://github.com/nim-lang/Nim/pull/15612).
+- `fusion` is now un-bundled from Nim, `./koch fusion` will
+  install it via Nimble at a fixed hash.
+- `testament`: added `nimoutFull: bool` spec to compare full output of compiler
+  instead of a subset; many bugfixes to testament.
+
+
+## Misc/cleanups
+- Deprecated `TaintedString` and `--taintmode`.
+- Deprecated `--nilseqs` which is now a noop.
+- Added `-d:nimStrictMode` in CI in several places to ensure code doesn't have certain hints/warnings.
+- Removed `.travis.yml`, `appveyor.yml.disabled`, `.github/workflows/ci.yml.disabled`.
+- `[skip ci]` now works in azure and CI pipelines, see detail in PR [#17561](https://github.com/nim-lang/Nim/pull/17561).
+
+
+# Footnotes
+Tested on a 2.3 GHz 8-Core Intel Core i9, 2019 macOS 11.5 with 64GB RAM.
+* [1] command used: `nim c -o:/tmp/main -d:danger --eval:'echo "hello world"'`
```

Comment:
```suggestion
* [1] command used: `nim c -o:/tmp/main -d:danger --eval:'echo "hello world"'`.
Note that the Nim result is without optimizing for size (via `--opt:size`), without stripping (`--passL:-s`), and without using full link-time optimization (`--passC:-flto`).
It's also fairly straightforward to create a statically linked 5 kB "hello world" binary via musl, and you can get down to ~150 bytes with some extra tricks.
```

I think it's worth adding a couple of sentences here, as binary size is often a focus in reddit/HN comment threads. Feel free to adapt.

I've avoided `-d:strip` and `-d:lto` here, as they don't currently work reliably on every platform IIRC (macOS?).

Reply:
@ee7 see commit `update "small binaries" section` ; 

I didnt' mention `and you can get down to ~150 bytes with some extra tricks.` as I don't see where are those extra tricks mentioned; in particular https://irclogs.nim-lang.org/07-07-2020.html#12:31:34 mentions:
> and if I disregard echo and use c_printf from system/ansi_c it's 5112 bytes

do you have more detail?

@Yardanico can you please confirm the 5K numbers for static linking with musl? on osx (without musl) i get the quoted 49K

Reply:
Sure, I'll check those numbers on Linux and Windows again to see how small will it be

Reply:
@timotheecour it's better to discuss this matter in real-time chats I think, but here are the results for Zig 0.8.1 + Nim 1.5.1 (6bb32da4aecbe80a9cc0cce90322c0e741ee8b85):
Nim source file:
```nim
echo "Hello, World!"
```

Compilation process:
```
$ nim c --os:any -d:posix -d:noSignalHandler --cc:clang --clang.exe="zigcc.sh" --clang.linkerexe="zigcc.sh" -d:danger --gc:arc -d:useMalloc --panics:on --passL:"-target x86_64-linux-musl -flto" --opt:size hello.nim
Hint: used config file '/home/dian/nim/config/nim.cfg' [Conf]
Hint: used config file '/home/dian/nim/config/config.nims' [Conf]
...................................................
Hint:  [Link]
Hint: gc: arc; opt: size; options: -d:danger
23293 lines; 0.350s; 25.414MiB peakmem; proj: /home/dian/test_bin_size/hello.nim; out: /home/dian/test_bin_size/hello [SuccessX]

$ strip -s -R .comment -R .note -R .note.ABI-tag hello

$ wc hello
  14  166 7296 hello
```

`zigcc.sh` is a shell script with the contents:
```sh
#!/usr/bin/env sh
zig cc $@
```
It's needed because Nim expects a name of the compiler binary and doesn't accept commands with spaces like `zig cc`.

I've also done a couple of tests with a few musl gcc toolchains, but the binary size with them is around 13KB even after stripping.

P.S.: Not sure if we should really specify the whole process for building a small binary, it'll be quite cumbersome to follow.

Reply:
Without the `--os:any -d:posix` trickery the file size is 16976 bytes.

Reply:
> Result - a fully static binary with the size of 7296 bytes.

@Yardanico I get 4848 bytes for a `echo "Hello, World!` binary on a similar setup (x86_64; Arch Linux; Nim 1.5.1 recent commit; Zig 0.8.1).

It looks like you forgot this option: `--passC:'-flto -target x86_64-linux-musl'`.

The same thing, as a script:

```sh
#!/usr/bin/env sh
set -e
path_bin='./hello'

# Create source file
path_src="${path_bin}.nim"
rm -f "${path_src}"
echo 'echo "Hello, World!"' > "${path_src}"

# Compile
nim c -f \
  -d:danger --opt:size --gc:arc --panics:on \
  -d:useMalloc --os:any -d:posix -d:noSignalHandler \
  --cc=clang --clang.exe='zigcc.sh' --clang.linkerexe='zigcc.sh' \
  --passC:'-flto -target x86_64-linux-musl' \
  --passL:'-flto -target x86_64-linux-musl' \
  --skipUserCfg --skipParentCfg --skipProjCfg \
  "${path_bin}"

# Strip
strip --strip-all --remove-section=.comment "${path_bin}"

# Print some information about the binary
printf "\n"
file "${path_bin}"
byte_count="$(wc --bytes ${path_bin} | cut -d' ' -f1)"
echo "${byte_count} bytes"

# Show that the executable actually works
printf "\nRunning executable...\n"
"${path_bin}"
```

```
./hello: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, stripped
4848 bytes

Running executable...
Hello, World!
```